[WIP] Split xxhsum benchmark code, use better timers

easyaspi314 opened the pull request:
- [x] Separate code into its own file
- [x] Add a `XSUM_NO_BENCH` option
- [ ] Cleanup
- [ ] Change benchmark to use the more accurate `timefn.c`

That last one is an issue — where should `timefn.c` go? 

I was thinking `/utils/` or `/common`. We could also put some of the functions we reuse there.

easyaspi314 committed:
xxhsum: Split out benchmark code

Benchmark code is now in cli/xsum_bench.c.

It is still a little rough.

easyaspi314 committed:
xxhsum: Add option for disabling benchmarks

Similar to XSUM_NO_TESTS, if the benchmark functions are called when
compiled with this option, it will print an error and exit 1.

If xxhsum.c was compiled with XSUM_NO_TESTS, the benchmark options will
be removed from the usage. However, only xsum_bench.c needs to be
compiled with this option.

Cyan4973 commented:
Good, this rewrite makes the code much clearer imho.

> where should `timefn.c` go?

Good question.

As a possible reference, in other repositories like `lz4` and `zstd`, 
I do have a strict include policy that enforces : `lib <- programs <- tests`,
aka `lib` must not depend on anything outside of `lib`,
`programs` can depend on `lib`, but nothing else.
A user should be able to extract _just_ `lib` and `programs`, and get it to compile `zstd` or `lz4` just fine.
Consequently, if there is something useful for `tests` hosted in `programs` (such as `timefn`),
`tests` has to include it from there.

Now, while this is strictly enforced in `lz4` and `zstd`, 
nothing like that exists yet in `xxhash` (since everything was flat so far).
To be fair, one can already note a few differences : 
we don't have a `lib` directory (`xxhash` lib is directly at root),
`programs` is replaced by `cli`,
and `tests` hosts each project separately into its own subfolder,
while `zstd` and `lz4` store many of them flat (mainly because there are multiple simple one-file test programs).

While we could go into other directions,
I presume it would be simpler for me if all repositories follow the same convention.
I have nothing against `common/` per se, it's a sensible choice,
just it would make `xxhash` repository a bit different in layout from `lz4` or `zstd`
(and lose the property that `cli/` depends only on root).
Not a big deal, it could go either way,
though that's exactly what conventions are for : arbitrate a choice, between otherwise equivalent options, to remain consistent and reduce surprises with other projects following the same conventions.

easyaspi314 commented:
So you are saying put `timefn.c` in `cli`?

Cyan4973 commented:
Yeah, let's start this way.

easyaspi314 committed:
Move timefn.c to cli/

Soon to be used in xxhsum benchmark

easyaspi314 committed:
timefn.c: add clock_gettime

Covers basically all relatively recent *nix platforms, and runs on the
same struct timespec that timespec_get does, allowing us to reuse the
codepath with little modification.

Easily detectable:

   #include <unistd.h>
   #if defined(_POSIX_TIMERS) && _POSIX_TIMERS > 0
   // have clock_gettime
   #endif

easyaspi314 committed:
xxhsum: use timefn.c in benchmark

More accurate, and reuses code we used elsewhere.

easyaspi314 commented:
Added `clock_gettime` to `timefn.c`: it is really easy to slip in and covers < C11. This means that output is deterministic on POSIX systems regardless of C11 support.

easyaspi314 committed:
timefn.c: silence some conversion warnings

easyaspi314 committed:
Fix conversion errors on Win32 path.

easyaspi314 committed:
Fix typo

PTime, not Ptime.

Cyan4973 commented on tests/bench/benchfn.c:
```diff
@@ -18,7 +18,7 @@
 #include <assert.h>      /* assert */
 
-#include "timefn.h"        /* UTIL_time_t, UTIL_getTime */
+#include <timefn.h>        /* UTIL_time_t, UTIL_getTime */
```
`timefn.h` is a project's header, not a system one.
It should use `""` double quotes

easyaspi314 replied:
Good point. I will fix that. I should just use relative paths, correct?

Cyan4973 replied:
correct

easyaspi314 committed:
Include sys/time.h, use quoted include for timefn.h

easyaspi314 committed:
Change benchfn.c to use relative paths

easyaspi314 committed:
timefn.h: Define _POSIX_C_SOURCE for struct timespec

Tricky one to find since I don't really use glibc.

Cyan4973 commented:
If you are still interested in this PR objective, I believe it would be a good next step.

Cyan4973 commented:
Well, a lot has changed since this PR was proposed,
in particular, all files related to the CLI, including `xxhsum.c`, are not hosted into the `cli` directory.
But I still believe that this PR has a good objective, so could be updated.

Cyan4973 commented:
Superseded by #614
